compute/cker/include/cker/train/operation/BinaryArithmetic.h

```diff
@@ -0,0 +1,99 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __NNFW_CKER_TRAIN_OPERATION_BINARYARITHMETIC_H__
+#define __NNFW_CKER_TRAIN_OPERATION_BINARYARITHMETIC_H__
+
+#include "cker/Shape.h"
+#include "cker/eigen/Utils.h"
+#include "cker/operation/BroadcastTo.h"
+
+namespace nnfw
+{
+namespace cker
+{
+namespace train
+{
+enum class ArithmeticType
+{
+  kAdd,
+  kSub,
+  kMul,
+  kDiv,
+};
+
+template <typename T>
+void BinaryArithmeticGrad(const Shape &lhs_shape, const T *lhs_data, const Shape &rhs_shape,
+                          const T *rhs_data, const Shape &incoming_shape, const T *incoming_data,
+                          const Shape &lhs_grad_shape, T *lhs_grad_data,
+                          const Shape &rhs_grad_shape, T *rhs_grad_data,
+                          ArithmeticType arithmetic_type)
+{
+  switch (arithmetic_type)
+  {
+    case ArithmeticType::kAdd:
+    {
+      BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), lhs_grad_shape, lhs_grad_data);
+      BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), rhs_grad_shape, rhs_grad_data);
+    }
+    break;
+
+    case ArithmeticType::kSub:
+    {
+      BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), lhs_grad_shape, lhs_grad_data);
+
+      auto const in_map = MapAsMatrixWithLastDimAsRows(incoming_data, incoming_shape);
+      auto rhs_grad_map = MapAsMatrixWithLastDimAsRows(rhs_grad_data, rhs_grad_shape);
+      rhs_grad_map = -in_map;
+    }
+    break;
+
+    case ArithmeticType::kMul:
+    {
+      auto const in_map = MapAsMatrixWithLastDimAsRows(incoming_data, incoming_shape);
+      auto const lhs_map = MapAsMatrixWithLastDimAsRows(lhs_data, lhs_shape);
+      auto const rhs_map = MapAsMatrixWithLastDimAsRows(rhs_data, rhs_shape);
+      auto lhs_grad_map = MapAsMatrixWithLastDimAsRows(lhs_grad_data, lhs_grad_shape);
+      auto rhs_grad_map = MapAsMatrixWithLastDimAsRows(rhs_grad_data, rhs_grad_shape);
+
+      lhs_grad_map = in_map.array() * rhs_map.array();
+      rhs_grad_map = in_map.array() * lhs_map.array();
+    }
+    break;
+
+    case ArithmeticType::kDiv:
+    {
+      auto const in_map = MapAsMatrixWithLastDimAsRows(incoming_data, incoming_shape);
+      auto const lhs_map = MapAsMatrixWithLastDimAsRows(lhs_data, lhs_shape);
+      auto const rhs_map = MapAsMatrixWithLastDimAsRows(rhs_data, rhs_shape);
+      auto lhs_grad_map = MapAsMatrixWithLastDimAsRows(lhs_grad_data, lhs_grad_shape);
+      auto rhs_grad_map = MapAsMatrixWithLastDimAsRows(rhs_grad_data, rhs_grad_shape);
+
+      lhs_grad_map = in_map.array() / rhs_map.array();
+      rhs_grad_map = in_map.array() * -lhs_map.array() / rhs_map.array() / rhs_map.array();
```

Comment:
I'm seeing weird outputs :/
Is there something wrong with the following?

$L$: LHS
$R$: RHS
$O$: Output (of division)
$X$: Output (of entire model)

$L / R = O$ (Elementwise division)

$\frac{\partial X}{\partial L} = \frac{\partial O}{\partial L} \cdot \frac{\partial X}{\partial O} = \frac{1}{R} \cdot \frac{\partial X}{\partial O}$
$\frac{\partial X}{\partial R} = \frac{\partial O}{\partial R} \cdot \frac{\partial X}{\partial O} = -\frac{L}{R^2} \cdot \frac{\partial X}{\partial O}$